datafusion/expr/src/type_coercion/binary.rs

```diff
@@ -777,9 +638,17 @@ fn is_time_with_valid_unit(datatype: DataType) -> bool {
 
 /// Coercion rules for Temporal columns: the type that both lhs and rhs can be
 /// casted to for the purpose of a date computation
-fn temporal_coercion(lhs_type: &DataType, rhs_type: &DataType) -> Option<DataType> {
+pub(crate) fn mathematics_temporal_common_type(
```

Comment:
It's for getting common type for temporal_type

Reply:
This is very good and makes a lot of sense. 

Given this logic is also used for comparisons of temporal types I think `mathematics_temporal_common_type` may be a misleading name I think leaving the name as `temporarl_coercion` and updating the docstrings might be the most readable

I also think it would be good to note in the comments that it doesn't handle the interval arithmetic (where the input types are different) only comparisons and arithmetic where the types are the same (like timestamp + timestamp)

Reply:
Eventually it would be great to support coercion for functions where the types don't have to be the same (e.g. https://github.com/apache/arrow-datafusion/issues/5652) but this is already much better than what we have today